### PR TITLE
Users myorders functionality 29

### PR DESCRIPTION
**Pull request by avneetmalhotra**

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -16,11 +19,35 @@ def destroy
   def show
   end
 
+  def myorders
+  end
+
+  def cancel
+    if @order.cancel
```
Comment (nimish13) [inline app/controllers/orders_controller.rb]: also add `canceller_id` in order table to keep a track who cancelled the order

File: app/models/concerns/checkout.rb [hunk 2]
```diff
@@ -7,13 +7,18 @@ module Checkout
...
+# delivered -> order has been delivered
 ## flow - cart -> address -> payment -> completed
 
   included do
     ## STATE MACHINE
     state_machine :state, initial: :cart do
       before_transition on: [:add_address, :pay, :complete], do: :checkout_allowed?
       after_transition on: :complete, do: [:set_completed_at, :decrease_deals_stock, :send_confirmation_instructions]
+      after_transition on: :deliver, do: :set_delivered_at
+      before_transition on: :cancel, do: :cancellation_allowed?
```
Comment (nimish13) [inline app/models/concerns/checkout.rb]: add this as if condition in transition

File: app/models/concerns/checkout.rb [hunk 3]
```diff
@@ -27,17 +32,29 @@ module Checkout
...
+      end
+
     end
   end
 
   def checkout_allowed?
     is_order_not_empty? && are_deals_live? && are_deals_quantity_valid?
   end
 
+  def cancellation_allowed?
```
Comment (nimish13) [inline app/models/concerns/checkout.rb]: make it as private if we are not using it outside the class

File: app/models/concerns/checkout.rb [hunk 4]
```diff
@@ -69,7 +94,11 @@ def is_order_not_empty?
 
     def are_deals_live?
       if deals.expired.present?
-        errors[:base] << I18n.t(:has_expired_deals, scope: [:order, :errors])
+        if completed?
```
Comment (nimish13) [inline app/models/concerns/checkout.rb]: please use a generic message

File: app/models/concerns/checkout.rb [hunk 5]
```diff
@@ -85,4 +114,24 @@ def are_deals_quantity_valid?
       end
     end
 
+    def can_be_cancelled?
+      if line_items.includes(:deal).all? { |line_item| line_item.deal.end_at - Time.current > ENV['minutes_before_expiration_when_deal_can_be_cancelled'].to_i.minutes }
```
Comment (nimish13) [inline app/models/concerns/checkout.rb]: check if this can be done via rails query

Comment (nimish13) [reply app/models/concerns/checkout.rb]: `ENV['minutes_before_expiration_when_deal_can_be_cancelled'].to_i.minutes` move this to a cosntant in constant.rb

File: app/models/line_item.rb [hunk 6]
```diff
@@ -42,6 +42,18 @@ def ensure_deal_not_bought_again_in_another_order
       end
     end
 
+    def ensure_deal_not_bought_again_in_another_order
```
Comment (nimish13) [inline app/models/line_item.rb]: please optimize

File: app/models/order.rb [hunk 7]
```diff
@@ -17,6 +17,9 @@ class Order < ApplicationRecord
   ## SCOPES
   scope :incomplete, ->{ where(completed_at: nil) }
   scope :complete, ->{ where.not(completed_at: nil) }
+  scope :open, ->{ where(state: 'completed') }
```
Comment (nimish13) [inline app/models/order.rb]: rename open => ready_for_delivery

File: app/views/orders/_order.html.erb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        <table class="table table-hover table-condensed">
+          <thead>
+            <th nowrap></th>
+            <th nowrap>TITLE</th>
+            <th nowrap>UNIT PRICE</th>
+            <th nowrap></th>
+
+          </thead>
+          <tbody>
+            <% order.line_items.each do |line_item| %>
```
Comment (nimish13) [inline app/views/orders/_order.html.erb]: avoid N+1 queries

File: app/views/orders/myorders.html.erb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+<div class="col-md-10 col-md-offset-1 col-xs-10 col-xs-offset-1">
```
Comment (nimish13) [inline app/views/orders/myorders.html.erb]: this should be index.html.erb

File: app/views/orders/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            <th nowrap>TITLE</th>
+            <th nowrap>UNIT PRICE</th>
+            <th nowrap>QUANTITY</th>
+            <th nowrap>LOYALTY DISCOUNT</th>
+            <th nowrap>TOTAL PRICE</th>
+            <th nowrap></th>
+
+          </thead>
+          <tbody>
+            <% @order.line_items.each do |line_item| %>
```
Comment (nimish13) [inline app/views/orders/show.html.erb]: avoid N+1 queries

**Comment by avneetmalhotra** [comment]
closed in reference to #33

**Pull request closed without merging**